core/src/Kokkos_MpiShmem.hpp

```diff
@@ -0,0 +1,223 @@
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact  H. Carter Edwards (hcedwar@sandia.gov)
+// 
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOS_MPISHMEM_HPP
+#define KOKKOS_MPISHMEM_HPP
+
+#include <Kokkos_Core_fwd.hpp>
+
+// If MPI execution space is enabled then use this header file.
+
+#if defined( KOKKOS_HAVE_MPISHMEM )
+
+#include <iosfwd>
+
+#include <Kokkos_MpiShmemSpace.hpp>
+
+#include <Kokkos_Parallel.hpp>
+#include <Kokkos_Layout.hpp>
+#include <Kokkos_ScratchSpace.hpp>
+#include <Kokkos_MemoryTraits.hpp>
+#include <impl/Kokkos_Tags.hpp>
+
+/*--------------------------------------------------------------------------*/
+
+namespace Kokkos {
+namespace Experimental {
+namespace Impl {
+class MpiShmemExec ;
+} // namespace Impl
+} // namespace Experimental
+} // namespace Kokkos
+
+/*--------------------------------------------------------------------------*/
+
+namespace Kokkos {
+namespace Experimental {
+
+/// \class MpiShmem
+/// \brief Kokkos Execution Space that uses MPI shared memory communicators.
+///
+/// An "execution space" represents a parallel execution model.  It tells Kokkos
+/// how to parallelize the execution of kernels in a parallel_for or
+/// parallel_reduce.  For example, the Threads execution space uses Pthreads or
+/// C++11 threads on a CPU, the OpenMP execution space uses the OpenMP language
+/// extensions, and the Serial execution space executes "parallel" kernels
+/// sequentially.  The MpiShmem execution space uses MPI's
+/// shared memory communicator splitting to execute kernels across MPI
+/// ranks on the same CPU.
+class MpiShmem {
+public:
+  //! \name Type declarations that all Kokkos execution spaces must provide.
+  //@{
+
+  //! Tag this class as a kokkos execution space
+  typedef MpiShmem              execution_space ;
+
+  //! This execution space's preferred memory space.
+  typedef MpiShmemSpace         memory_space ;
+
+  //! This execution space preferred device_type
+  typedef Kokkos::Device<execution_space,memory_space> device_type;
+
+  //! The size_type best suited for this execution space.
+  typedef memory_space::size_type  size_type ;
+
+  //! This execution space's preferred array layout.
+  typedef LayoutRight            array_layout ;
+
+  //!
+  typedef ScratchMemorySpace< MpiShmem >  scratch_memory_space ;
+
+  //@}
+  //--------------------------------------------------
+  //! \name Functions that all Kokkos devices must implement.
+  //@{
+
+  /// \brief True if and only if this method is being called in a
+  ///   thread-parallel function.
+  inline static int in_parallel() {
```

Comment:
If you're using MPI + MPI with processes at both levels, and MPI has been initialized, the code is _always_ in parallel.  Thus, `in_parallel()` should return the result of `MPI_Initialized()`.